### DISCUSSION TITLE
How to keep data persistent across multiple connections without interfering with each other

### DISCUSSION THREAD

**Discussion by bsdcpp** [mitmproxy/mitmproxy#5670] [General]
Hi, everyone:
    I want to maintain a variable between multiple connections of a device, because there will be multiple devices and multiple connections, so the variables of multiple devices cannot be overwritten with each other.
   I don't know if there is a function similar to "object-oriented programming", every device gets an instance.
   I am currently trying to implement it with mutex lock, but I don't know how to get the device's information the connection come from.

```python
from multiprocessing import Lock
class func:
    def __init__(self):
        self.mutex = Lock()
        self.meta = {}
        pass

    def client_connected(self, client):
        self.mutex.acquire()
        self.device_info = client.peername[0]
        self.mutex.release()
        return

    def set_kv(self, info):
        self.mutex.acquire()
        self.meta[self.device_info] = info
        self.mutex.release()
        return

    def get_kv(self):
        self.mutex.acquire()
        ret = self.meta[self.device_info]
        self.mutex.release()
        return ret

    def request(self, flow: mitmproxy.http.HTTPFlow):
        url = flow.request.url

        if 'xxx' in url:
            var_get_from_url = "url...."
            self.set_kv(var_get_from_url)

        if 'yyy' in url:
            func(self.get_kv())
        return
```

There is still a problem here, that is, getting the device info from client_connected , and setting value  for the device. not a 
 atomic operation .
    It would be nice if the connected device information could be obtained directly from the request event.
Thanks.

**Comment by bsdcpp** [comment]
Through dir(flow), I found that there is a client_conn method, which temporarily solved the problem. I don't know if there is a better way to make the data persistent.
```python
from multiprocessing import Lock
class func:
    def __init__(self):
        self.mutex = Lock()
        self.meta = {}
        pass

    def set_kv(self, dev, info):
        self.meta[dev] = info
        return

    def get_kv(self, dev):
        ret = self.meta[dev]
        return ret

    def request(self, flow: mitmproxy.http.HTTPFlow):
        url = flow.request.url

        if 'xxx' in url:
            self.mutex.acquire()
            dev = flow.client_conn.peername[0]
            var_get_from_url = "url...."
            self.set_kv(dev, var_get_from_url)
            self.mutex.release()

        if 'yyy' in url:
            self.mutex.acquire()
            dev = flow.client_conn.peername[0]
            var_get_from_url = "url...."
            func(self.get_kv(dev))
            self.mutex.release()

        return
```

**Comment by Prinzhorn** [comment]
Without the `@concurrent` decorator all hooks run in the same thread. There is no need for a mutex. But even with `@concurrent` you don't need a mutex, as `flow.client_conn.peername` can be considered immutable (for a given flow instance the connection won't be changed).

The way you are currently using the mutex doesn't make sense. There is a single instance of your `wfkids` addon. But its methods will be called for all connections. So your `self.device_info` will change for every new connection but you still get request/response events for an entirely different connection.

Each connection has a unique `client_conn.id`. So I would map the id to the meta info you need and remember to clear it when the connection closes or else you leak memory with each connection. Unless of course you want it per device IP and not each connection, then keep using what you're using.

**Reply by bsdcpp** [reply]
Thank you very much for your answer, @concurrent decorator you mentioned helped me a lot. Better than a mutex.

**Reply by mhils** [reply]
For the record, I would strongly discourage the use of `@concurrent` unless you exactly know what you are doing. mitmproxy's proxying logic is always single-threaded (`@concurrent` does not change that), so you can only expect performance gains if your hooks are highly CPU intensive or if you need to work with blocking I/O APIs. Using `async` hooks is generally much better. You also don't need any locking then.
